gcc 6.1.0

zbeekman opened the pull request:
Update gcc formula for 6.1 release. Passes tests locally, and brew audit --strict etc.

AFAICT ISL, cloog, etc. are all fine as is.

zbeekman commented:
I had another commit with also resolved the suffix issue when `--HEAD` is passed, but I couldn't get the `brew test --HEAD` to pass... Is there an equivalent to `build.head?` for the test block?

DomT4 commented:
> Is there an equivalent to build.head? for the test block?

There is, simply `if head?`, but we don't really provide support for testing `HEAD`; it's too difficult to formally support due to the lack of any stability.

DomT4 commented:
There's a bunch of things that will need revisions here and potentially tests rewriting looking at the CI failures.

zbeekman commented:
@DomT4... let me take a look at the failures

zbeekman commented:
It appears things are hardwired to use gcc-5? @DomT4 yes it looks like someone hard-wired a ton of stuff to gcc-5

DomT4 commented:
A lot of things actually link to some of the GCC libraries, and with the suffix change from 5 to 6 here those will need rebuilding so they can "find" the GCC libraries again and ensure they'll actually build okay against the new GCC. A lot of that will be linkage with the `fortran` libraries.

zbeekman commented:
@DomT4 yes that is certainly true... I never really understood why package maintainers added the suffixes to begin with. It just seems to be more hassle than it's worth.

Also, GCC transitioned to a new release schedule where a new major release will come out about every 12 months... which if every time you upgrade GCC you break half your system (because everything is used to looking for 5 instead of 6) that seems like a giant PITA. I have yet to comb through the release notes in great detail (other than some Fortran features) so I can't say for certain that there won't be any backwards compatibility issues.

Could you point me to where the tests are defined?

Also, does it make sense to upgrade the `gcc` formula now? Or would it make sense to add a `gcc6` formula?

dunn commented:
There already is a gcc6 formula in Versions, which might be a helpful reference: https://github.com/Homebrew/homebrew-versions/blob/master/gcc6.rb

Would it be worth adding an option for the Go compiler (`gccgo`)?

UniqMartin commented:
Just to add to the list of things that need to be done: We also need to adjust [`compilers.rb`](https://github.com/Homebrew/brew/blob/9378e46ffc2c17df4efc535744279ebfa28bcf30/Library/Homebrew/compilers.rb#L3-L4) and possibly other pieces of Homebrew code to accept the new version.

DomT4 commented:
> I never really understood why package maintainers added the suffixes to begin with.

Well, on OS X at least there's already a `gcc` and if we overwrote that in the `$PATH` you could expect quite some trouble 😄.

> Also, GCC transitioned to a new release schedule where a new major release will come out about every 12 months... which if every time you upgrade GCC you break half your system (because everything is used to looking for 5 instead of 6) that seems like a giant PITA. 

This is why we force rebuilds on things. I don't think breakage every 12 months that we can handle in advance and should be more-or-less invisible to users is that much of a problem compared to something like Boost which breaks with laughable regularity in often annoyingly subtle ways. I don't see anything massively breaking in the changelog, but I'm not a compiler expert _(and actively want to avoid becoming one at all and any cost)_.

> Could you point me to where the tests are defined?

Which tests?

> Also, does it make sense to upgrade the gcc formula now? Or would it make sense to add a gcc6 formula?

It makes sense to at least try here. If everything builds and tests pass, we're good.

> We also need to adjust compilers.rb and possibly other pieces of Homebrew code to accept the new version.

👍.

zbeekman commented:
@DomT4: 
> Which tests?

The tests that brew bot runs. It appears a lot of them are hard-wired for gcc-5... 

@dunn: Thanks for the reference... I'll compare these formulae to scrutinize for other subtle changes that may be required which I missed. I'm not sure why gcc6 exists, since there is an option for gcc with --HEAD now... but its helpful for my current purposes

@UniqMartin: now that brew and core are separate, how should we manage the PRs? I might be able to find time to take a stab at it...

DomT4 commented:
> The tests that brew bot runs. It appears a lot of them are hard-wired for gcc-5...

The `linkage` ones are a developer command ran by CI to spot broken links, i.e. if something previously  used `libfoo5.dylib` and now `libfoo5.dylib` is missing there's a problem and in this case the formula needs its `revision` bumped.

The `test` ones are the result of failures running `brew test --verbose <formula>` where `<formula>` is something that needs `gcc`.

Almost all of them look like your standard "dylib version has changed, revision to see if that fixes the problem" dance.

UniqMartin commented:
> now that brew and core are separate, how should we manage the PRs? I might be able to find time to take a stab at it...

It makes sense to handle the stuff in `Homebrew/brew` first, as it's mostly about recognizing the new version when users do something as esoteric as `brew install --cc=gcc-6` and for properly dealing with `fails_with :gcc => "6"` stanzas in formulae. Making those parts recognize GCC 6 before it is available as a formula shouldn't break anything. (You might want to have a look at the Git history to find out what other parts were touched when support for GCC 5 was added.) Thanks!

zbeekman commented:
@DomT4: Ah, got it now. Here's my plan of action:

1. scrutinize gcc6 from versions to make sure I'm not missing something subtle
2. @dunn what should I do about gccgo? Should I build it by default, with option to disable or add option to build, but disable by default?
3. I'll adjust the --HEAD install to allow brew test --HEAD gcc to work... even though you don't require it
4. I'll do my best to figure out what needs adjusting by looking at what happened when gcc-5 was added... will the history be preserved across the new repo layout?
5. Open PR against brew to get internals ready for gcc-6
6. Bump formula revisions to relink them with gcc-6 and resolve linkage issues: @DomT4 how should I break up the commits? Should I include the revision bumps in the same commit as gcc-6 version bump?

@all how does this plan of action sound?

dunn commented:
> what should I do about gccgo? Should I build it by default, with option to disable or add option to build, but disable by default?

Probably off by default?  I don't really know anything about it beyond that it exists now.

DomT4 commented:
> I'll do my best to figure out what needs adjusting by looking at what happened when gcc-5 was added... will the history be preserved across the new repo layout?

It's a bit split. Core changes will have been applied to `homebrew/brew` and formulae changes to this repo now. It might be easier in this case if you just hunt down the GCC 5 PR from `homebrew/homebrew-legacy` for comparison.

> @DomT4 how should I break up the commits? Should I include the revision bumps in the same commit as gcc-6 version bump?

Same PR, but one commit per formula revision with the commit message being something like:
```
foo: bump revision for gcc
```

UniqMartin commented:
> 1. I'll adjust the --HEAD install to allow brew test --HEAD gcc to work... even though you don't require it

Don't waste your time on that. We tend to reject those test additions, just like with code that tests non-standard options. Tests are mostly for our CI use …

> 1. I'll do my best to figure out what needs adjusting by looking at what happened when gcc-5 was added... will the history be preserved across the new repo layout?
> 2. Open PR against brew to get internals ready for gcc-6

It should be as simple as Homebrew/legacy-homebrew#39470, just with even fewer changes.

zbeekman commented:
@UniqMartin: Just opened up [a PR](https://github.com/Homebrew/brew/issues/163) following https://github.com/Homebrew/legacy-homebrew/pull/39470

@DomT4: Following procedure in Homebrew/legacy-homebrew#39905

zbeekman commented:
Also, devil, dwarf and yap all fail with GCC 5, which I'm assuming means that they will fail with GCC 6. I can update the `fails_with` section, but I don't think these require a revision bump, right?

UniqMartin commented:
> Also, devil, dwarf and yap all fail with GCC 5, which I'm assuming means that they will fail with GCC 6. I can update the `fails_with` section, but I don't think these require a revision bump, right?

It would be nice to check whether that's still the case, but if it is, updating the `fails_with` is sufficient and no revision bump is needed.

DomT4 commented:
> Following procedure in Homebrew/legacy-homebrew#39905

Note that due to the new `linkage` checks being more comprehensive in terms of not relying on a good formula test to work there's likely more formulae that'll need fixing this time.

zbeekman commented:
@DomT4 I think they are all enumerated in the tests, no? Any way I'll try to slog through it

DomT4 commented:
There shouldn't be anything extra that needs either revision bumping that isn't named [here](http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/1055/#showFailuresLink). Whether stuff fails to build after being revision bumped remains to be seen; that's the really fun part.

zbeekman commented:
That's why fails_with exists right? lol j/k ;-)

On Thu, Apr 28, 2016 at 2:22 PM Dominyk Tiller <notifications@github.com>
wrote:

> There shouldn't be anything extra that needs either revision bumping that
> isn't named here
> <http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/1055/#showFailuresLink>.
> Whether stuff fails to build after being revision bumped remains to be
> seen; that's the really fun part.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Homebrew/homebrew-core/pull/676#issuecomment-215517950>
>

zbeekman commented:
crazy day... wasn't able to work on this much... also, it looks like a regression in gcc 6.1 will affect at least one of the packages that needs a formula revision bump: json-fortran (I am the formula maintainer/contributor and work on the project as a volunteer... upstream is aware of the issue and has a work around that should be in a future release)

zbeekman commented:
sorry for the silence here... I had a work trip and then a bunch of fires to put out when I returned... carving out some time for this now...

zbeekman commented:
Does anyone know if [the last patch in gcc.rb](https://github.com/Homebrew/homebrew-core/blame/master/Formula/gcc.rb#L264-L282) is needed in gcc 6.1? It seems as though it is not present in the gcc6 homebrew-versions formula

zbeekman commented:
Also, the gcc6.rb formula seems to claim that the [java issues](https://github.com/Homebrew/homebrew-core/blame/master/Formula/gcc.rb#L176-L186) have also been properly resolved.

zbeekman commented:
@dunn it appears gccgo is still unsupported on OS X; it raises an error for unsupported language during configure

zbeekman commented:
I'm testing locally again, and pushing some minor changes, including trusting the gcc6.rb on the java sandboxing issues, because it appears the brew-bot/Jenkins logs get deleted after a while and I need to see which formula versions I need to bump to resolve linkage and other errors

zbeekman commented on Formula/gcc.rb:
```diff
@@ -172,19 +174,7 @@ def install
-        add_suffix file, version_suffix if File.exist? file
-      end
-    end
+    # Since GCC 4.9 java properties are properly sandboxed.
```
I grabbed this hunk and comment from homebrew-versions/gcc6.rb

Is there a way to test whether this can be safely removed? Should I just keep it in?

CC: @UniqMartin @DomT4 @mikemcquaid @dunn

mikemcquaid replied:
I'm not sure. Could check `git blame` and see who added it?

zbeekman replied:
It [looks like](https://github.com/Homebrew/homebrew-core/blame/master/Formula/gcc.rb#L176-L186) it was first added by you ( @mikemcquaid ) June 6th, 2014 to gcc.rb in https://github.com/Homebrew/homebrew-core/commit/9cc81cee476e5d956afa2f0d58820200848a2fb1

@manphiz [removed it](https://github.com/Homebrew/homebrew-versions/blame/master/gcc6.rb#L180) in the initial addition of homebrew-versions/gcc6.rb back in January 2016

mikemcquaid replied:
It was actually added before then; I was reverting another commit. What's the path of `logging.properties` after this is installed with `--with-java`?

zbeekman replied:
Still looks like it was added by you 😛 in https://github.com/Homebrew/homebrew-core/commit/59aeccfbab2bb958ed5435760733411b9bd78955 ... I'm building `--with-java` right now so I can see where GCC tries to install the files in question.

zbeekman replied:
@mikemcquaid It looks like they are installed in the same location: lib/gcc/6/ ... but the specialized code above is renaming them from e.g., `lib/gcc/6/logging.properties` to `lib/gcc/6/logging-6.properties`

```
$ find /usr/local/homebrew/Cellar/gcc/6.1.0 -name 'logging.properties'
/usr/local/homebrew/Cellar/gcc/6.1.0/lib/gcc/6/i386/logging.properties
/usr/local/homebrew/Cellar/gcc/6.1.0/lib/gcc/6/logging.properties
```

here is the add suffix function:

```ruby
  def add_suffix(file, suffix)
    dir = File.dirname(file)
    ext = File.extname(file)
    base = File.basename(file, ext)
    File.rename file, "#{dir}/#{base}-#{suffix}#{ext}"
  end
```

@manphiz do you know why this was done in gcc6.rb?

mikemcquaid replied:
@zbeekman I'm guessing you can link multiple GCC 6 versions at once. At a guess this was a problem for GCC 4 versions but given the new numbering may not be.

zbeekman replied:
@mikemcquaid Thanks for getting back so quickly. I'll leave this hunk alone unless someone directs me to revert it or gives me more specific tests to run.

mikemcquaid replied:
Looking at https://github.com/Homebrew/homebrew-versions/blob/master/gcc6.rb I think you can remove this now; this is never going to clash with another formula on major/minor versions, it seems.

zbeekman replied:
> Looking at https://github.com/Homebrew/homebrew-versions/blob/master/gcc6.rb I think you can remove this now; this is never going to clash with another formula on major/minor versions, it seems.

:+1:

zbeekman commented on Formula/gcc.rb:
```diff
@@ -33,6 +33,8 @@ def osmajor
   end
 
+  # GCC's Go compiler is not currently supported on Mac OS X.
+  # See: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=46986
```
@dunn it seems that go still won't work on OS X... the GCC configure script complains about unsupported language if you try to enable it

dunn replied:
Got it, thanks for investigating!

zbeekman commented on Formula/json-fortran.rb:
```diff
@@ -1,8 +1,8 @@
-  url "https://github.com/jacobwilliams/json-fortran/archive/4.3.0.tar.gz"
-  sha256 "c97f8de53e2ca9ee91fb3148bfa971b00eaea6e757b5e3d67cc4b4ff197e392e"
+  url "https://github.com/jacobwilliams/json-fortran/archive/5.0.0.tar.gz"
+  sha256 "0a1bf8788bdf8bdc6af009ae078ed374b2c95e0cfc3f30354b5dc5d3d35e5e66"
```
JSON-Fortran 4.3 triggered a bug in GCC 6.1 (gfortran) that was not triggered in 5.3. A fix to this bug was included in the latest 5.x release.

We are considering back porting to a 4.4 release since 5.0.0 is not backwards compatible, but provides many improvements and bug fixes

zbeekman replied:
I may split this out  into a separate PR, in hopes that  it can get merged before this PR

zbeekman commented:
Could someone please let me know what the deal is with brew-bot/jenkins? It seems that it has just been sitting here for well over 12 hours, and I'd like to be able to look at the output to determine which remaining formulas need revision bumps or more involved triage

CC: @UniqMartin @DomT4 @mikemcquaid

mikemcquaid commented:
@BrewTestBot test this please

zbeekman commented on Formula/grass.rb:
```diff
@@ -1,7 +1,7 @@
   desc "Geographic Resources Analysis Support System"
   homepage "https://grass.osgeo.org/"
-  revision 1
+  revision 2
```
This  formula is  missing a `test do` block. It appears to be a GUI tool (built from source) and if adding a `test do` block is even feasible, I won't be able to do it.

zbeekman commented on Formula/abi-compliance-checker.rb:
```diff
@@ -3,6 +3,7 @@ class AbiComplianceChecker < Formula
   homepage "http://ispras.linuxbase.org/index.php/ABI_compliance_checker"
   url "https://github.com/lvc/abi-compliance-checker/archive/1.99.9.tar.gz"
   sha256 "52b0daca89fcda73cde126497c8015ca823417074ba02fcff68b7acf2f45e516"
+  revision 1
```
This is still failing, even after the revision bump: http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/1959/version=mavericks/console

dunn replied:
It failed in that case just because Jenkins timed out.  I think Mike bumped the timeout for the build that's happening right now.

DomT4 replied:
IIRC the test needs rewriting to account for GCC switching to single-digit versioning, `gcc-5` instead of `gcc-4.9` for example. This has been a problem for a while but has been stuck on various to-do lists.

zbeekman replied:
yes that's what it looks like... I'll poke around for a few minutes and see if I can determine how to fix this...

zbeekman replied:
easy fix.... will push once I go through test results

DomT4 commented:
> It seems that it has just been sitting here for well over 12 hours, and I'd like to be able to look at the output to determine which remaining formulas need revision bumps or more involved triage

Apologies, I pushed you off the queue yesterday because the queue was already a touch on the too long side and this PR will take a while to test.

zbeekman commented:
@DomT4 no worries....

I have some questions for everyone:

 1. Can we up the time limit for the tests on this PR? I don't know if I've finished bumping formula revisions etc. unless all the tests run
 1. I have fixed the  abi-compliance-checker test and am ready to push, but should I wait until someone can extend/disable the brew-bot timeouts for this PR?
 1. grass.rb is going to fail due to missing `test do` block. It is beyond my capabilities to add one. Will this be a problem for building bottles, etc?
 1. Should I break out JSON-Fortran 5.0.0 into it's own PR? 5.0.0 includes a bug work around for GCC 6.1 so it should be merged before this PR, if I break it into a separate one...

DomT4 commented:
> Can we up the time limit for the tests on this PR? I don't know if I've finished bumping formula revisions etc. unless all the tests run

I bumped the timeout to two hours, for now. It's worth noting that Jenkins doesn't offer a per-PR timeout, sadly.

It sort of has to come down as soon as possible because inevitably someone files a PR with some sort of bug in the build or test and everyone else sits around waiting for the timeout because the maintainers aren't on station 24/7 😄.

> grass.rb is going to fail due to missing test do block. It is beyond my capabilities to add one. Will this be a problem for building bottles, etc?

It shouldn't, actually. `brew audit --strict` only runs on new formulae, and `brew audit` vanilla doesn't flag missing tests.

> Should I break out JSON-Fortran 5.0.0 into it's own PR? 5.0.0 includes a bug work around for GCC 6.1 so it should be merged before this PR, if I break it into a separate one...

Either seems fine to me, but if you don't include the version bump here it's likely you'll need to include a revision bump here.

zbeekman commented:
> I bumped the timeout to two hours, for now. It's worth noting that Jenkins doesn't offer a per-PR timeout, sadly.

@DomT4 we're 20 minutes away from hitting the 2 hour time limit and still building bottles... any way you could increase the time limit while the build is occurring?

zbeekman commented:
It appears that only one more formula needs a revision bump: gasnet.rb... but sam2p and grass are failing to build. I looked into sam2p and couldn't really tell what is going on. I tried building it locally, but still no dice, although my system is not as pristine as the testbot is... here is a `brew git-logs sam2p` from my local machine: https://gist.github.com/zbeekman/bbf4230a867a1500c80febf56eb5c9e6 Specifically take a look at configure.log.

I am trying to build grass locally now, to see what's going on.

dunn commented:
If Grass is a blocker we can boneyard it, since our version is well out of date and the @Homebrew/geospatial tap should be up and running soon.

zbeekman commented:
@dunn yes right now grass and sam2p are blockers. I have tried to build grass locally and posted logs to gist: https://gist.github.com/b88fc63aaac9644f2e6f25eb6fe26f07

I'm seeing the same errors reported by brew-bot:
```
[snip]
configure:895: gcc-4.2 -o conftest    conftest.c  1>&5
<built-in>:0: error: Unknown value '10.10' of -mmacosx-version-min
configure: failed program was:

#line 890 "configure"
#include "confdefs.h"

main(){return(0);}
```

dakcarto commented:
@dunn wrote:
> If Grass is a blocker we can boneyard it, since our version is well out of date and the @Homebrew/geospatial tap should be up and running soon.

Yes, we are starting the process of prepping the formulae in https://github.com/OSGeo/homebrew-osgeo4mac for transition to @Homebrew/geospatial.

Also, our `grass-64` formulae already compiles with `clang`.

Note, we will not be maintaining the versioned current formulae names we use in osgeo4mac in Homebrew/geospatial, only for older formulae that need to stick around, and excepting maybe were the version difference is significant, like `gdal` and `gdal2`. So, our GRASS 6 formula will go into Homebrew/geospatial as `grass6` and GRASS 7 will be `grass`. This is because GRASS 6 is still an optional external tool for QGIS.

zbeekman commented:
Proposal: boneyard grass and sam2p, and bump gasnet revision. I think that should conclude stuff that needs addressing, save, perhaps, a few formulae that have `fails_with` for gcc 5 that might also fail with gcc 6...

After that I think this PR should be pretty close to finished. I may need to do some commit splitting and re-working as I didn't quite follow @DomT4 's [instructions](https://github.com/Homebrew/homebrew-core/pull/676#issuecomment-215479622) (sorry, got lost in the noise)

DomT4 commented:
> we're 20 minutes away from hitting the 2 hour time limit and still building bottles... any way you could increase the time limit while the build is occurring?

Nope. Once it's set and the job starts you either cancel the job and queue a new one with an even bigger timeout or you wait and hope. Never let it be said that CI solutions are flexible 😉.

zbeekman commented:
@DomT4 it squeaked through with a few to spare. Any indication of the current job is stuck or is there just a big queue right now?

dunn commented:
I guess technically we're migrating `grass` to the geospatial tap instead of boneyarding it.  Can you update `tap_migrations.json` for it and sam2p?
```json
"grass": "homebrew/geospatial",
...
"sam2p": "homebrew/boneyard"

DomT4 commented:
The latter. Apparently the world chose tonight to file a bunch of relatively heavy builds across multiple taps. It's currently doing `swift` which will take a while.

dunn commented:
Actually we can add that when we pull to avoid running it through CI again.

DomT4 commented:
> I guess technically we're migrating grass to the geospatial tap instead of boneyarding it.

Are we? That tap doesn't seem to exist yet (?), and it'll need to as on `update` people who currently have `grass` installed will be migrated to the new tap.

dakcarto commented:
@DomT4 wrote:
> Are we? That tap doesn't seem to exist yet (?), and it'll need to as on update people who currently have grass installed will be migrated to the new tap.

Indeed. `Homebrew/homebrew-geospatial` repo _needs created_ and assigned to the @Homebrew/geospatial team. Then I can start requesting others to be added to that team. I don't appear to have permissions to create the new repo.

Also, I would prefer to not migrate the older `grass` formula, but instead move the newer `grass-64` one out of osgeo4mac to `homebrew-geospatial/Formula/grass6.rb`.

DomT4 commented:
^ https://github.com/Homebrew/homebrew-geospatial. Can't add/remove new people from that team yet, I'll probably leave that to Mike to sign off on in the morning, but you should be able to make a start on adding Formula inside a `Formula` subdirectory as desired.

dakcarto commented:
@DomT4, thanks!

dunn commented:
Thanks @DomT4!  Sorry @dakcarto, I didn't realize it hadn't been set up.

mikemcquaid commented:
@dakcarto The previously discussed plan was for you to migrate the existing tap to the Homebrew organisation (which you have/had rights to do). That seems like a better fit than creating a new tap from scratch, no? If you agree, I'll delete the new repository and you can transfer the existing one.

DomT4 commented:
I'd presumed the request for the new tap was deliberate given how explicit it was despite prior discussions so I created it. My apologies if that was a misunderstanding in the end.

mikemcquaid commented:
@DomT4 no worries, not your fault.

zbeekman commented:
OK, so everything is good except for GASNet. There is a new release of GASNet, and I'm going to see if I can also sort out some issues it was having with UDP/SMP conduit, since MPI is somewhat of a dumb choice to use on most Macs... hopefully I can kill 3 birds with one stone here (upgrade, fix brew-bot errors, sort out UDP/SMP stuff reported by Jeff Hammond

zbeekman commented:
Proposal: boneyard GASNet for the time being. The issue is that the GASNet installation creates makefile stubs that are supposed to be included by the user. However, they end up pointing to compilers and linkers in `/usr/local/homebrew/.../ENV/4.3/...` which don't work outside of the homebrew build process. Ideally these should be patched so point to the compiler and linker that the scripts in `.../ENV/4.3` resolve to.

zbeekman commented:
@DomT4: I have cleaned up my commits and formatted them according to [your guidance](https://github.com/Homebrew/homebrew-core/pull/676#issuecomment-215479622).

@dunn I have also edited `tap_migrations.json` according to [your guidance](https://github.com/Homebrew/homebrew-core/pull/676#issuecomment-219874234).

@mikemcquaid @DomT4 @UniqMartin This rebase should trigger a build that passes so long as it doesn't time out. Could you please ensure that brew-bot/jenkins has a timeout of 2.5 hours? If the build needs to be re-triggered to take advantage of the time out, can you please re-trigger it?

If this build passes, I'm feeling pretty comfortable merging it. The areas that may need more attention from all of you are:

 1. [`tap_migrations.json`](https://github.com/Homebrew/homebrew-core/pull/676/files#diff-5d7c9de419327e968bff35664a993971) : I don't know how the migration mechanics work... this also goes hand-in-hand with the next point:
  1. migrations of the actual formulae: I have put in [a PR 120 against homebrew-boneyard](https://github.com/Homebrew/homebrew-boneyard/pull/120) but I have no idea if this is required....
  1. What to do about homebrew-geospatial and `grass.rb` Right now I have just deleted `grass.rb` from core and added an entry in `tap_migrations.json` pointing it to homebrew-geospatial

dakcarto commented:
@mikemcquaid wrote:
> @dakcarto The previously discussed plan was for you to migrate the existing tap to the Homebrew organisation (which you have/had rights to do). That seems like a better fit than creating a new tap from scratch, no? If you agree, I'll delete the new repository and you can transfer the existing one.

Hmm, that might be the best solution now that I think about it. It will take a bit longer to prepare our osgeo4mac repo for migration, since I need to revamp all the formulae (remove versioned formula names, do strict audit, etc.) first.  

Yes, please remove the current homebrew-geospatial repo and its one PR.

mikemcquaid commented:
> Hmm, that might be the best solution now that I think about it. It will take a bit longer to prepare our osgeo4mac repo for migration, since I need to revamp all the formulae (remove versioned formula names, do strict audit, etc.) first.

@dakcarto You don't need to do any of that. Can we move it first and then tweak it afterwards? Have been waiting for a while for this. Thanks!

zbeekman commented:
it seems @ilovezfs aborted the build.... not sure why

dakcarto commented:
@mikemcquaid wrote:
> @dakcarto You don't need to do any of that. Can we move it first and then tweak it afterwards? Have been waiting for a while for this. Thanks!

Yeah, sorry about the long wait. My past employment took me far away from Homebrew activities for a while. This also means the tap is not currently in great shape, and it's not very Homebrew-standard-like. would prefer to clean it up before moving. I can probably do so within a week, with further tweaking after it is moved and (hopefully) configured with `brew-test-bot`.

On a plus note, my new employer may be able to fund my work on the homebrew-geospatial tap in the near future.

We should move this conversation out of this ticket, as there is more to discuss (what formulae to migrate from other homebrew-* taps, versioning, etc.). Should I make a new issue at homebrew-core, or elsewhere?

mikemcquaid commented:
@dakcarto It's OK for it to not be in great shape or a Homebrew standard; I'd just really like to get it moved in ASAP so I can fix up the permissions, get you CI etc. If it's just going to be a week we can wait but technically nothing needs to change (including URLs as people will be redirected) if you move to the Homebrew org.

zbeekman commented:
build #2020 will fail because I forgot to include open-mpi.rb revision bump in refactored commits.. it can be canceled...

zbeekman commented:
@BrewTestBot test this please

zbeekman commented:
@ilovezfs @DomT4 @UniqMartin @mikemcquaid I don't know why but the last test tried to checkout a nonexistent commit

zbeekman commented:
Alright folks, I'm pretty pleased with how things stand. I don't know if I have addressed the boneyard migrations 100% correctly, and please also see PR https://github.com/Homebrew/homebrew-boneyard/pull/120

As for grass, I have removed it and added the migration to homebrew-geospatial to the tap_migrations.json file, but right now AFAIK there is no geospatial tap with a grass.rb in it... so that probably needs to be addressed.

As for GASNet, the formula was more or less never functional... I'm going to try to submit a new formula soon to fix this, but for now I migrated it to boneyard... it may not even warrant saving in boneyard, idk, someone who knows what they're doing should weigh in here...

So pending final code review from @DomT4 @mikemcquaid or someone else, I think this is good to go. :shipit:

mikemcquaid commented:
We need to wait for the homebrew-geospatial tap to exist so this PR is currently blocked on @dakcarto. Otherwise it looks good to me.

dakcarto commented:
@mikemcquaid wrote:
> We need to wait for the homebrew-geospatial tap to exist so this PR is currently blocked on @dakcarto. Otherwise it looks good to me.

Ok, I'll work on migrating the homebrew-osgeo4mac tap to homebrew-geospatial today. Hopefully have it ready by tonight or tomorrow.

Only blocking issue for migration is `gdal`, which I would like to see in `homebrew-geospatial` as `gdal` (v2.0.2, then moving to 2.1), with `gdal111` (v1.11.x) in homebrew-geospatial or homebrew-versions (there is an outdated one in homebrew-versions right now).  I'm dropping all versioned formula in the osgeo4mac tap in favor of just versioning for older versions needing to stick around, so the scheme is more in-line with other Homebrew formulae.

Which means, something similar to this issue still needs addressed for formulae dependent upon `gdal`: https://github.com/Homebrew/legacy-homebrew/pull/42845

Formulae in osgeo4mac tap (QGIS, GRASS, SAGA) already use osgeo4mac's `gdal-20` formula. I _could_ move that over as `gdal2`, but would prefer not to version the current formula, only outdated (but still needed) ones.

Alternatively, the `gdal` in homebrew-core could be upgraded to 2.0.2 (or 2.1) first, with non-compatible formulae using `gdal111`.

Recommendations?

dunn commented:
@dakcarto we can rename formulae now, so settling on names isn't critical.

dakcarto commented:
@dunn wrote:
> @dakcarto we can rename formulae now, so settling on names isn't critical.
>  Show all checks

By "now," what do you mean. Has this changed recently, beyond aliases, like in past 6-12 months? Please elaborate on the process.

mikemcquaid commented:
@dakcarto https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Rename-A-Formula.md The main caveat is that the name may not be "freed up" after renaming.

zbeekman commented:
Hi everyone, just curious if there's been any movement on homebrew-geospatial... I may split out e83e795f205ca4fb51b88f2488878a1319adba1b into its own PR if this is going to take a while....

UniqMartin commented:
Not sure about the state of `homebrew/geospatial` but splitting out the version bump of `json-fortran` into a separate PR is reasonable, if the new version works equally fine with GCC 5.3. Ultimately, splitting out changes that can be done independently of the GCC version bump (because they work with both the 5.3 and 6.1 version) is a good thing because it helps to reduce the complexity of this PR.

dakcarto commented:
@zbeekman wrote
> Hi everyone, just curious if there's been any movement on homebrew-geospatial... I may split out e83e795 into its own PR if this is going to take a while....

Yes, I worked on updating the formulae in osgeo4mac (~50% done), in preparation for the migration to homebrew-geospatial. Nothing committed yet, as I want all the changes to hit at once, then immediately migrate the tap. I'm currently at a QGIS conference, and should be able to pick it back up early next week.

zbeekman commented:
@UniqMartin yup, totally backwards compatible *with GCC 5.3* (It is a 5.0.0 release with major new features and functionality so it is **not** backwards compatible with JSON-Fortran 4.x.x .... I would like to back port the GCC 6 fixes to 4.x branch, and perhaps make available through homebrew versions?)

@dakcarto Thanks for the update.

UniqMartin commented:
> I would like to back port the GCC 6 fixes to 4.x branch, and perhaps make available through homebrew versions?

While possible in theory, I doubt this formula is popular enough. It currently has less than one download per day (if bottle downloads are an indicator of popularity). We try to keep the focus of `homebrew/versions` limited to formulae where there is high demand for legacy versions and are currently in the process of purging as many of the existing formulae in there as possible. Most of them are just being dumped in that tap and then immediately stop being cared for, effectively causing more harm than good.

zbeekman commented:
@UniqMartin fair. Would it make more sense, then, to try to publish a 4.x release *first* that is GCC 6.1 compatible, with a new formula, and then merge the 5.0.0 formula some time later? (So anyone who wants GCC 6.1 but doesn't want to upgrade to 5.0.0 with breaking changes can keep using 4.x? by pinning?)

UniqMartin commented:
I think getting out 5.0.0 out to users now is the best option, as Homebrew users tend to prefer to live on the bleeding edge most of the time. If there turns out to be sufficient demand for the 4.x series, we can still add a `json-fortran4` formula in `homebrew/versions` or possibly in a separate tap maintained by you.

zbeekman commented:
OK sounds good, and makes my life easier!

zbeekman commented:
@UniqMartin: sorry, one more question, if I pull out JSON-Fortran 5.0.0 into it's own PR, should I re-push after rebasing? (This will trigger a CI rebuild, which will fail unless the time limit is extended and unless I rebase onto the merged PR for JSON-Fortran will also cause a linkage failure)

UniqMartin commented:
Yes, but please wait with pushing until the `homebrew/geospatial` situation has been resolved one way or another and the updated `json-fortran` formula has been merged, so you can rebase on top of that.

Looking at the diff, I think it's also worthwhile to split out boneyarding `gasnet` and `sam2p` into individual PRs. This will give them better visibility in the issue tracker and will allow us to handle these earlier and as individual tasks. Also means less stuff that needs to be done in the correct order (coordinating with the `homebrew/boneyard` PRs) for whoever ends up merging this PR.

zbeekman commented:
> Yes, but please wait with pushing until the homebrew/geospatial situation has been resolved one way or another and the updated json-fortran formula has been merged, so you can rebase on top of that.

👍 
 
> Looking at the diff, I think it's also worthwhile to split out boneyarding gasnet and sam2p into individual PRs. This will give them better visibility in the issue tracker and will allow us to handle these earlier and as individual tasks. Also means less stuff that needs to be done in the correct order (coordinating with the homebrew/boneyard PRs) for whoever ends up merging this PR.

OK. I have a PR open against homebrew-boneyard right now. I'll split out the boneyard commits, and reference the new PR for those from the boneyard PR.

zbeekman commented:
@UniqMartin I've split out the boneyard and JSON-Fortran update into #1368 and #1369 Please let me know if there's anything on those PRs that you would like me to address.

UniqMartin commented:
@zbeekman Another minor thing that would be nice to address before the next push to this PR's branch would be to slightly clean up the commit subjects of some commits. For revision bump commits we prefer the commit subject to be `<formula>: revision for gcc` (we tend to omit the version here). However, currently I'm seeing three different commit subject formats used in the relevant commits in your branch and none of them matches our preferred style. Thanks!

zbeekman commented:
Yup, happy to do it... I'll include this in the rebase after JSON-Fortran is sorted out and homebrew-geospatial comes into existence

zbeekman commented:
Any word on Homebrew geospatial? (I think that is the last issue blocking this, other than force pushing my cleaned up commits, without JSON-Fortran, and without boneyard migrations...)

zbeekman commented:
The tests will probably time out, but it might be worthwhile to extend the time limit and re-run them, since I've rebased these commits and re-worded according to @UniqMartin's suggestion, and dropped commits taken care of elsewhere... I think this is just blocked by grass/homebrew-geospatial

UniqMartin commented:
I aborted the build, bumped the timeout to three hours, and restarted the build. What's missing from the commits is a revision bump for `json-fortran`. It has been version bumped in a separate PR (as I advised), but that bottle was built against libraries from GCC 5, so with GCC 6 entering the scene it will need to be rebuilt against the new libraries in the new locations (`libgfortran` et al.).

Not sure what we should do about `grass`. If the `homebrew/geospatial` tap isn't going to materialize soon, I can basically think of two options:

- Temporarily boneyard `grass` and all formulae depending on it.
- Make the `grass` formula to compile with Clang – seems to be already done according to https://github.com/Homebrew/homebrew-core/pull/676#issuecomment-219836523 – and make the necessary changes in the core `grass` formula. It can be migrated later once the new tap is ready.

zbeekman commented:
ah yes... I forgot that JSON-Fortran also needs a revision bump... thanks for the reminder.

Just pushed the JSON-Fortran revision...

As far as `grass` goes, I'm happy with either approach, I'll let others decide what's best.

dakcarto commented:
@zbeekman, I will update the `grass` formula to compile with clang. Please do not migrate it to `homebrew-geospatial` as there will be other `grass` formulae migrated there via a repo migration instead (which I am still working on).

zbeekman commented:
@dakcarto why not just temporarily boneyard it? It sounds like `homebrew-geospatial` will be the preferred location to get grass, and it will no longer be in core, so what's the advantage of porting it to clang, only to have that then superseded by the other version of `grass` in `homebrew-geospatial`?

If you do have a good reason to port to build with clang, how should be proceed? I could remove 31f07b3 via interactive rebase, and then force push to zbeekman:master again. Then you could open a PR against zbeekman:master? Or I guess just open a PR against homebrew core, let them merge that, then I can rebase on top of that?

MikeMcQuaid commented:
@zbeekman Let's boneyard it. With all due respect to @dakcarto I don't think it's fair for the GCC upgrade to be blocked on his work again.

zbeekman commented:
@MikeMcQuaid Should I do the boneyard in a separate PR, or here?

MikeMcQuaid commented:
@zbeekman Whichever you prefer.

UniqMartin commented:
Let's do this in a separate PR because it involves multiple formulae:

- `grass` itself.
- `gdal-grass` which has a hard dependency on `grass`.
- `homebrew/science/qgis` that optionally depends on `grass` (and that we might decide to ignore).

zbeekman commented:
What needs to happen for dal-grass ad qgis? does the dependency just change to `boneyard/grass`?

UniqMartin commented:
`gdal-grass` needs to be boneyarded together with `grass` as it has a hard dependency and we don't allow for hard cross-tap dependencies in core (and depending on formulae in `homebrew/boneyard` is completely forbidden). This can happen in the same PR (and a matching PR that adds them in them to the [`homebrew/boneyard` tap repository](https://github.com/Homebrew/homebrew-boneyard)).

`qgis` needs to be stripped of its optional `grass` dependency together with everything that references this dependency in the formula code and submitted as separate PR to the [`homebrew/science` tap repository](https://github.com/Homebrew/homebrew-science).

UniqMartin commented:
Pulled the PR and bottles from the last CI run. Dropped the conflicting commit (`grass` migration) locally as it doesn't have an influence on the build and there's no point in going through CI one more time.

Merged in 27d2f81e20b79aca95019f7c460a14b4a6e24183. Thank you for another contribution to Homebrew, @zbeekman! And a massive thanks for pulling through this non-trivial version bump and the related complications. It has been quite a ride. You rock! :tada:

zbeekman commented:
Thanks to you folks at Homebrew for all your hard work! Cheers @UniqMartin @DomT4 @dunn @MikeMcQuaid and everyone else!

zbeekman commented:
@sjackman does this get pulled into linuxbrew automatically?

sjackman commented:
Hi, Izaak. Yes, it will be pulled into Linuxbrew eventually. I'm behind the nominal weekly merge schedule right now due to work deadlines. Hopefully in the next week you can expect GCC 6 in Linuxbrew.
